posts/2018-08-16-well-typed-heap.lhs

```diff
@@ -0,0 +1,1100 @@
+inductive proofs (we're using `lemma1` and `lemma2` here).  We don't go in too
+much detail with our explanation here -- this code is often hard to write but
+surprisingly easy to read.
+
+> lumberjack_go _ FEnd = VNull
+> lumberjack_go nnat0 (F0 forest0) = fmap
+>     (\cutTree -> case cutTree of
+>         CutTree xnat (SSucc nnat) t1 forest1 -> CutTree
+>             (SSucc xnat)
+>             nnat
+>             (case lemma2 nnat xnat of QED -> t1)
+>             (F0 forest1))
+>     (lumberjack_go (SSucc nnat0) forest0)
+> lumberjack_go nnat0 (F1 tree0 forest0) = VCons
+>     (CutTree
+>         SZero
+>         nnat0
+>         (case lemma1 nnat0 of QED -> tree0)
+>         (F0 forest0))
+>     (fmap
+>         (\cutTree -> case cutTree of
+>             CutTree xnat (SSucc nnat) t1 forest1 -> CutTree
+>                 (SSucc xnat)
+>                 nnat
+>                 (case lemma2 nnat xnat of QED -> t1)
+>                 (F1 tree0 forest1))
+>         (lumberjack_go (SSucc nnat0) forest0))
+
+Lumberjack: final form
+----------------------
+
+Now that we can select `Popcount b` trees, it time to convert this to something
+more convenient to work it.  We will use a `NonEmpty` to represent our list of
+candidates to select from.
+
+> lumberjack
+>     :: forall b a. BNonZero b ~ 'True
+>     => Forest 'Zero b a
+>     -> NonEmpty.NonEmpty (CutTree 'Zero b a)
+
+
+First we select the `Popcount b` trees:
+
+> lumberjack trees =
+>     let cutTrees :: Vec (Popcount b) (CutTree 'Zero b a)
+>         cutTrees = lumberjack_go SZero trees in
+
+Then we convert it to a `NonEmpty`.  This requires us to call `lemma3` (the
+proof that relates non-zeroness of a binary number with non-zeroness of a
+natural number through popcount).  We need an appropriate `SBin` to call
+`lemma3` and the auxiliary function `treesToSBin` defined just below does that
+for us.
+
+>     case lemma3 (treesToSBin trees :: SBin b) of
+>          QED -> vecToNonEmpty cutTrees
+
+This sneaky function constructs an `SBin b` from a `Forest o b a` value.  It is
+maybe a bit naughty but this prevents us from carrying around the singletons in
+a bunch of places.
+
+> treesToSBin :: Forest o b a -> SBin b
+> treesToSBin FEnd     = SBEnd
+> treesToSBin (F0 t)   = SB0 (treesToSBin t)
+> treesToSBin (F1 _ t) = SB1 (treesToSBin t)
+
+popHeap: gluing the pieces together
+-----------------------------------
+
+We can now find all trees in the heap that may be cut.  They are returned in a
+`CutTree` datatype.  If we assume that we are taking a specific `CutTree`, we
+can take the root from the tree inside this datatype, and we can construct a new
+heap from its children using `childrenToForest`.  Then, we merge it back
+together with the original heap.
+
+The new heap has one less element -- hence we use `BDec` (binary decrement,
+define just a bit below).
+
+> popForest
+>     :: forall a b. Ord a
+>     => CutTree 'Zero b a
+>     -> (a, Forest 'Zero (BDec b) a)
```

Comment:
is it really necessary to define `BDec`? In this case, you can write:

```
popForest :: forall a b. Ord a => CutTree 'zero (BInc b) a -> (a, Forest 'Zero b a)
```

Reply:
You can do it with just `BInc` but you end up with a more awkward signature for `popHeap` involving `Width`, and type inference isn't as good for things like `:t popHeap h` in GHCi, so I'll leave it as is.